activesupport/lib/active_support/cache.rb

```diff
@@ -781,31 +774,35 @@ def dup_value!
       end
 
       private
-        def should_compress?(options)
-          if @value && options.fetch(:compress, true)
-            compress_threshold = options.fetch(:compress_threshold, DEFAULT_COMPRESS_LIMIT)
-            serialized_value_size = (@value.is_a?(String) ? @value : marshaled_value).bytesize
+        def compress!(compress_threshold)
+          case @value
+          when nil, true, false, Numeric
+            uncompressed_size = 0
+          when String
+            uncompressed_size = @value.bytesize
+          else
+            serialized = Marshal.dump(@value)
+            uncompressed_size = serialized.bytesize
+          end
+
+          if uncompressed_size >= compress_threshold
+            serialized ||= Marshal.dump(@value)
+            compressed = Zlib::Deflate.deflate(serialized)
 
-            serialized_value_size >= compress_threshold
+            if compressed.bytesize < uncompressed_size
+              @value = compressed
+              @compressed = true
+            end
           end
         end
 
         def compressed?
-          defined?(@compressed) ? @compressed : false
-        end
-
-        def compress!
-          @value = Zlib::Deflate.deflate(marshaled_value)
-          @compressed = true
+          defined?(@compressed)
```

Comment:
Yes. Ruby semantics of query methods don't require us to return `false` or `true` so `nil` is totally fine and inline with our guidelines.

Reply:
@rafaelfranca 
Ruby semantics is fine. I have another concern here since I see that behavior of this method is changed here.
`defined?` returns a string describing its argument by http://ruby-doc.org/core-2.5.0/doc/keywords_rdoc.html
Example:
```ruby
@compressed = nil # or @compressed = false
defined? @compressed # => "instance-variable"
```
That's mean if we used `@compressed` like `nil` or `false`, we would need to change this method to the previous version in order to get right behavior.
I would recomend to initialize `@compressed` in the controller in order to get rid of using `defined?` there.
if it is reasonoble for you, I'll send a patch?
```diff
diff --git a/activesupport/lib/active_support/cache.rb b/activesupport/lib/active_support/cache.rb
index d769e2c..317723b 100644
--- a/activesupport/lib/active_support/cache.rb
+++ b/activesupport/lib/active_support/cache.rb
@@ -719,6 +719,7 @@ def initialize(value, compress: true, compress_threshold: DEFAULT_COMPRESS_LIMIT
         @version    = version
         @created_at = Time.now.to_f
         @expires_in = expires_in && expires_in.to_f
+        @compressed = nil

         compress!(compress_threshold) if compress
       end
@@ -798,7 +799,7 @@ def compress!(compress_threshold)
         end

         def compressed?
-          defined?(@compressed)
+          @compressed
         end

         def uncompress(value)
```

Reply:
The current implementation is fine. A string is a truthy value and that is only want we need. `@compressed` will never be `@nil` or `@false` because we serialize this instance using marshal and we don't want to always serialize `@compressed` and your implementation always serialize `@compressed`.

Reply:
You are right. Sorry for distraction.

Reply:
No problem! thank you for double checking the impementation.